PowerEditor/src/MISC/Common/Common.cpp

```diff
@@ -1329,25 +1329,112 @@ int nbDigitsFromNbLines(size_t nbLines)
 	return nbDigits;
 }
 
+namespace
+{
+	constexpr TCHAR timeFmtEscapeChar = 0x1;
+	constexpr TCHAR middayFormat[] = _T("tt");
+
+	// Returns AM/PM string defined by the system locale for the specified time.
+	// This string may be empty or customized.
+	generic_string getMiddayString(const TCHAR* localeName, const SYSTEMTIME& st)
+	{
+		generic_string midday;
+		// Obtain the length of the resulting string first.
+		int ret = GetTimeFormatEx(localeName, 0, &st, middayFormat, nullptr, 0);
+		if (ret > 0)
+		{
+			midday.resize(ret);
+			ret = GetTimeFormatEx(localeName, 0, &st, middayFormat, &midday[0], static_cast<int>(midday.size()));
+			if (ret != 0)

```

Comment:
Better to check `if (ret > 0)` here, too?

Reply:
`GetTimeFormatEx` returns 0 (in case of error) or the length of the resulting string. 
The error is unlikely here because we check the length first at line 1343. But I check it nevertheless. 
Empty AM/PM string is handled here as well (it will result in `midday.resize(0)`). 
`std::string` works not very well in some cases if it contains `'\0'`.

Reply:
Yes, both `GetTimeFormatEx` and `GetDateFormatEx` return 0 in error case and should never ever return negative values. To be on the safest side one can be, a check for `if (ret > 0)` after every call of the methods would even cover this. But it's just a minor remark.

Another question. Can it be avoided to call `GetTimeFormatEx` two times, because it looks like overhead? Maybe like this, or is there a downside?

```
constexpr int size = MAX_PATH;
generic_string midday;
midday.resize(size);

int ret = GetTimeFormatEx(localeName, 0, &st, middayFormat, &midday[0], size);
if (ret > 0)
{
    midday.resize(ret - 1); // Remove the null-terminator.
}
else
{
    midday.resize(0); // Clear.
}

return midday;
```